scalafix-nsc/src/main/scala/scalafix/nsc/NscSemanticApi.scala

```diff
@@ -11,110 +11,466 @@ import scalafix.ScalafixConfig
+              }
+              if (typeSymbol.exists) {
+                enclosingScope.enterIfNew(typeSymbol)
+                addRename(typeSymbol, to)
+              }
+            case _ =>
+          }
+          super.traverse(t)
+        case _ => super.traverse(t)
       }
-      gtree match {
-        case g.PackageDef(pid, _) =>
-          val newCtx = ctx.copy(enclosingPackage = pid.symbol.fullName + ".",
-                                inScope = ctx.inScope ++ members(pid.tpe))
-          gtree.children.foldLeft(newCtx)(evaluate)
-          ctx // leaving pkg scope
-        case t: g.Template =>
-          val newCtx =
-            ctx.copy(inScope = t.symbol.owner.fullName :: ctx.inScope)
-          gtree.children.foldLeft(newCtx)(evaluate)
-          ctx
-        case g.Import(expr, selectors) =>
-          val newNames: Seq[String] = selectors.collect {
-            case g.ImportSelector(from, _, to, _) if from == to =>
-              Seq(s"${expr.symbol.fullName}.$from")
-            case g.ImportSelector(_, _, null, _) =>
-              members(expr.tpe)
-          }.flatten
-          ctx.copy(inScope = ctx.inScope ++ newNames)
-        case _ =>
-          gtree.children.foldLeft(ctx)(evaluate)
+    }
+  }
+
+  private class OffsetTraverser extends g.Traverser {
+    val offsets = mutable.Map[Int, g.Tree]()
+    val treeOwners = mutable.Map[Int, g.Tree]()
+    override def traverse(t: g.Tree): Unit = {
+      t match {
+        case g.ValDef(_, name, tpt, _) if tpt.nonEmpty =>
+          offsets += (tpt.pos.point -> tpt)
+          treeOwners += (t.pos.point -> t)
+        case g.DefDef(_, _, _, _, tpt, _) =>
+          offsets += (tpt.pos.point -> tpt)
+          treeOwners += (t.pos.point -> t)
+        case _ => super.traverse(t)
       }
     }
-    evaluate(SemanticContext("", Nil), gtree)
-    builder
   }
 
   private def getSemanticApi(unit: g.CompilationUnit,
                              config: ScalafixConfig): SemanticApi = {
-    val offsets = offsetToType(unit.body, config.dialect)
     if (!g.settings.Yrangepos.value) {
-      val instructions = "Please re-compile with the scalac option -Yrangepos enabled"
-      val explanation  = "This option is necessary for the semantic API to function"
+      val instructions =
+        "Please re-compile with the scalac option -Yrangepos enabled"
+      val explanation =
+        "This option is necessary for the semantic API to function"
       sys.error(s"$instructions. $explanation")
     }
 
+    // Compute scopes for global and local imports
+    val st = new ScopeTraverser(unit)
+    st.traverseScopes()
+    val rootPkg = st.topLevelPkg
+    val rootImported = st.scopes(rootPkg)
+    val realRootScope = rootPkg.info.members
+
+    // Compute offsets for the whole compilation unit
+    val traverser = new OffsetTraverser
+    traverser.traverse(unit.body)
+
+    def toMetaType(tp: g.Tree) = {
+      config.dialect(tp.toString).parse[m.Type].get
```

Comment:
I'm not sure if the type converter will solve the problem either. Kind-projector synthesizes unique type names for each appearance of `Type.Name(?)`. The synthesized types look like this

```scala
cats.laws.discipline.CartesianTests.Isomorphisms[[β$0$]cats.free.Free[Option,β$0$]]
```

I'm not 100% sure, but I think those should be pretty-printed as `?`. Maybe we'll need to special case kind-projector since it's a very popular compiler plugin.

The best solution for now is to gracefully handle the parse error and not shorten that type signature. The type signature of `shortenType` indicates that it's always a safe operation, and indeed, you can always choose not to shorten the type.